M2/Macaulay2/d/boost-regex.cpp

```diff
@@ -0,0 +1,290 @@
+  //  optimize
+  collate = (1 << 8), /* make [a-b] locale sensitive */
+
+  /* flags for Perl and POSIX */
+  //  newline_alt
+  //  no_except: this is always on
+  //  save_subexpression_location
+
+  /* flags for Perl */
+  no_mod_m = (1 << 12), /* don't match ^ $ with newlines */
+  no_mod_s = (1 << 13), /* don't match . with newlines */
+  //  mod_s
+  //  mod_x
+  //  no_empty_expressions
+
+  /* flags for POSIX ERE */
+  no_escape_in_lists = (1 << 17), /* disable \ escapes in lists */
+  no_bk_refs = (1 << 18),         /* disable backreferences */
+
+  /* The rest are based on standard Boost match flag types. */
+  // https://www.boost.org/doc/libs/release/libs/regex/doc/html/boost_regex/ref/match_flag_type.html
+  //  match_not_bob
+  //  match_not_eob
+  //  match_not_bol
+  //  match_not_eol
+  //  match_not_bow
+  //  match_not_eow
+  match_any = (1 << 25),        /* return any match */
+  match_not_null = (1 << 26),   /* match must be nonempty */
+  match_continuous = (1 << 27), /* match must start at the beginning */
+  //  match_partial
+  //  match_single_line
+  match_prev_avail = (1 << 30),      /* --first is a valid iterator position */
+  match_not_dot_newline = (1 << 31), /* doesn't match . with newlines */
+  //  match_not_dot_null
+  //  match_posix
+  //  match_perl
+  //  match_nosubs
+  //  match_extra
+
+  //  format_sed
+  //  format_perl
+  //  format_literal
+
+  //  format_no_copy
+  //  format_first_only
+  //  format_all
+};
+
+/* returns a compile state machine for regex */
+boost::regex regex_compile(const M2_string pattern, const long flags)
+{
+  typedef std::pair<int /* pattern_hash */, long /* flags */> ExpressionKey;
+  static std::map<ExpressionKey, boost::regex> cache;
+
+  /* attempt to recover a compiled expression from a map */
+  auto key = ExpressionKey {strings_hash(pattern), flags};
+  auto hit = cache.find(key);
+  if (hit != cache.end()) return hit->second;
+#if DEBUG_REGEX & 1
+  std::cerr << "flags:\t" << flags << std::endl
+            << "regexp:\t" << M2_tocharstar(pattern) << std::endl;
+#endif
+
+  /* parse and set the flags */
+  boost::regex_constants::syntax_option_type regex_flags =
+      boost::regex::no_except; /* never throw exceptions */
+  regex_flags |= flags & extended ? boost::regex::extended : 0;
+  regex_flags |= flags & literal ? boost::regex::literal : 0;
+  regex_flags |= flags & icase ? boost::regex::icase : 0;
+  regex_flags |= flags & nosubs ? boost::regex::nosubs : 0;
+  regex_flags |= flags & collate ? boost::regex::collate : 0;
+  regex_flags |= flags & no_mod_m ? boost::regex::no_mod_m : 0;
+  regex_flags |= flags & no_mod_s ? boost::regex::no_mod_s : 0;
+  /* the following are on by default on POSIX ERE, so we disable them */
+  regex_flags &= flags & no_escape_in_lists ? regex_flags
+                                            : ~boost::regex::no_escape_in_lists;
+  regex_flags &= flags & no_bk_refs ? regex_flags : ~boost::regex::no_bk_refs;
```

Comment:
It's probably a mistake to complement these, as "no" is part of both names.

Reply:
It's not. As the comment line says, these two are enabled by **default** in `extended`, so we need to complement them, since they would introduce breaking changes from GNU regex.

Reply:
Shall I mark this as resolved? As I explained, the two flags are included in `regex::extended` by default, so we remove them to match the behavior of GNU regex.